### PR TITLE
Sema: Explicit stricter minimization of protocol composition types for redeclaration checking

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Enable explicit stronger minimization of protocol compositions via generic signature computation for purposes like redeclaration checking.

Note: This is a source-breaking change, because the following is now a redeclaration error:
```swift
protocol P: class {}
func foo(_: P) {}
func foo(_: P & AnyObject) {}
```

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@slavapestov What if we special-case protocol composition types in `ASTMangler::getDeclTypeForMangling` to preserve the mangling and make this ABI-compatible?

**Comment by slavapestov** [comment]
I'd prefer not to add any special cases and avoid changing getCanonicalType() to minimize protocol compositions. If you need to minimize protocol compositions for something, you can add a new method to ProtocolCompositionType (or ExistentialLayout or whatever) which performs it explicitly.

**Comment by AnthonyLatsis** [comment]
Are we ready to take on the source break if I make stricter minimization explicit and use it in redeclaration checking, as well as the other part of this PR (https://github.com/apple/swift/pull/39694/commits/4602bf114edf5c6117bca4ed830649377fb45a68)?

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@slavapestov Could I ask for another opinion on this? I kept it narrow as per your suggestion.

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -3744,11 +3792,72 @@ Type ProtocolCompositionType::get(const ASTContext &C,
...
+
+  // Use generic signature minimization: the requirements of the signature will
+  // represent the minimal composition.
+  auto *GParam = GenericTypeParamType::get(/*isTypeSequence=*/false, 0, 0, Ctx);
+  const auto Sig = buildGenericSignature(
+      Ctx, nullptr, {GParam},
+      {Requirement(RequirementKind::Conformance, GParam, CanTy)});
+  if (!Sig) {
+    return CanTy;
+  }
```
Comment (slavapestov) [inline lib/AST/Type.cpp:3834]: Cool! Can this use ASTContext::getOpenedArchetypeSignature()?

**Review by slavapestov — APPROVED** [review]

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@slavapestov Much obliged

**Pull request merged**